mass-deployer/pkg/mass-deployer/deployer.go

```diff
@@ -0,0 +1,243 @@
+	err := tfPluginClient.NetworkDeployer.BatchDeploy(ctx, deployments.networkDeployments)
+	if err != nil {
+		cancelContractsOfFailedDeployments(tfPluginClient, deployments.networkDeployments, []*workloads.Deployment{})
+		return []string{}, err
+	}
+
+	err = tfPluginClient.DeploymentDeployer.BatchDeploy(ctx, deployments.vmDeployments)
+	if err != nil {
+		cancelContractsOfFailedDeployments(tfPluginClient, deployments.networkDeployments, deployments.vmDeployments)
+		return []string{}, err
+	}
+	vmsInfo := loadDeploymentsInfo(tfPluginClient, deployments.deploymentsInfo)
+
+	return vmsInfo, nil
+}
+
+func buildDeployments(vms []Vms, nodesIDs []int, sshKeys map[string]string) groupDeploymentsInfo {
+	var vmDeployments []*workloads.Deployment
+	var networkDeployments []*workloads.ZNet
+	var deploymentsInfo []vmDeploymentInfo
+	nodesIDsIdx := 0
+
+	// here we loop over all groups of vms within the same node group, and for every group
+	// we loop over all it's vms and create network and vm deployment for it
+	// the nodesIDsIdx is a counter used to get nodeID to be able to distribute load over all nodes
+	for _, vmGroup := range vms {
+		for i := 0; i < int(vmGroup.Count); i++ {
+			nodeID := uint32(nodesIDs[nodesIDsIdx])
+			nodesIDsIdx = (nodesIDsIdx + 1) % len(nodesIDs)
+
+			disks, mounts := parseDisks(vmGroup.Name, vmGroup.SSDDisks)
+
+			network := workloads.ZNet{
+				Name:        fmt.Sprintf("%s%dnetwork", vmGroup.Name, i),
+				Description: "network for mass deployment",
+				Nodes:       []uint32{nodeID},
+				IPRange: gridtypes.NewIPNet(net.IPNet{
+					IP:   net.IPv4(10, 20, 0, 0),
+					Mask: net.CIDRMask(16, 32),
+				}),
+				AddWGAccess: false,
+			}
+			w := workloads.VM{
+				Name:        fmt.Sprintf("%s%d", vmGroup.Name, i),
+				NetworkName: network.Name,
+				Flist:       vmGroup.Flist,
+				CPU:         int(vmGroup.FreeCPU),
+				Memory:      int(vmGroup.FreeMRU),
+				PublicIP:    vmGroup.PublicIP4,
+				PublicIP6:   vmGroup.PublicIP6,
+				Planetary:   vmGroup.Planetary,
+				RootfsSize:  int(vmGroup.Rootsize * 1024), // Rootsize is in MB
+				Entrypoint:  vmGroup.Entrypoint,
+				EnvVars:     map[string]string{"SSH_KEY": sshKeys[vmGroup.SSHKey]},
+				Mounts:      mounts,
+			}
+			deployment := workloads.NewDeployment(w.Name, nodeID, "", nil, network.Name, disks, nil, []workloads.VM{w}, nil)
+
+			vmDeployments = append(vmDeployments, &deployment)
+			networkDeployments = append(networkDeployments, &network)
+			deploymentsInfo = append(deploymentsInfo, vmDeploymentInfo{nodeID: nodeID, deploymentName: deployment.Name, vmName: w.Name})
+		}
+	}
+	return groupDeploymentsInfo{vmDeployments: vmDeployments, networkDeployments: networkDeployments, deploymentsInfo: deploymentsInfo}
+}
+
+func cancelContractsOfFailedDeployments(tfPluginClient deployer.TFPluginClient, networkDeployments []*workloads.ZNet, vmDeployments []*workloads.Deployment) {
```

Comment:
this function actually cancels working deployments as the contract won't be set in the workloads unless it managed to deploy fine. and using after failed batch deployments make the whole process of deploying a batch of networks and vms start from the beginning instead of retying only the failed ones.